core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java

```diff
@@ -107,6 +118,45 @@ public boolean isEnabled() {
       return enabled;
    }
 
+   @Override
+   public void handleInMemoryExpiration(InternalCacheEntry<K, V> entry) {
+      dataContainer.compute(entry.getKey(), ((k, oldEntry, factory) -> {
+         if (entry == oldEntry) {
+            // We have to delete from shared stores as well to make sure there are not multiple expiration events
+            persistenceManager.deleteFromAllStores(k, PersistenceManager.AccessMode.BOTH);
+            if (cacheNotifier != null) {
+               cacheNotifier.notifyCacheEntryExpired(k, entry.getValue(), entry.getMetadata());
+            }
+            return null;
+         }
+         return oldEntry;
+      }));
+   }
+
+   @Override
+   public void handleInStoreExpiration(K key) {
+      dataContainer.compute(key, ((k, oldEntry, factory) -> {
+         if (oldEntry == null || (oldEntry.canExpire() && oldEntry.isExpired(timeService.time()))) {
```

Comment:
Yes now that I think about it there is a small gap between when the reaper thread found an expired entry before it acquire the data container lock that if another thread removes the value during that time it would inadvertently raise the event.  I am not sure how I can fix this other than rereading the store again which is pretty ugly, or have the store acquire the lock during the purging (even worse).